source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+            xReturn = eAzureIoTProvisioningFailed;
+        }
+        else if( ( xResult = AzureIoTMQTT_Init( &( xAzureIoTProvisioningClientHandle->_internal.xMQTTContext ),
+                                                pxTransportInterface, prvGetTimeMillseconds, prvEventCallback,
+                                                pucBuffer, ulBufferLength ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Provisioning initialization failed: with mqtt init error : %d", xResult ) );
+            xReturn = eAzureIoTProvisioningInitFailed;
+        }
+        else
+        {
+            xReturn = eAzureIoTProvisioningSuccess;
+        }
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+void AzureIoTProvisioningClient_Deinit( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
+{
+    if( xAzureIoTProvisioningClientHandle == NULL )
+    {
+        AZLogError( ( "AzureIoTProvisioningClient_Deinit failed: Invalid argument" ) );
+    }
+    else
+    {
+        xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponsePayload = NULL;
+        xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponseTopic = NULL;
+    }
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTProvisioningClientResult_t AzureIoTProvisioningClient_SetSymmetricKey( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle,
+                                                                               const uint8_t * pucSymmetricKey,
+                                                                               uint32_t ulSymmetricKeyLength,
+                                                                               AzureIoTGetHMACFunc_t xHmacFunction )
+{
+    AzureIoTProvisioningClientResult_t xReturn;
+
+    if( ( xAzureIoTProvisioningClientHandle == NULL ) ||
+        ( pucSymmetricKey == NULL ) || ( ulSymmetricKeyLength == 0 ) ||
+        ( xHmacFunction == NULL ) )
+    {
+        AZLogError( ( "Provisioning client symmetric key fail: Invalid argument" ) );
+        xReturn = eAzureIoTProvisioningInvalidArgument;
+    }
+    else
+    {
+        xAzureIoTProvisioningClientHandle->_internal.pucProvisioningSymmetricKey = pucSymmetricKey;
+        xAzureIoTProvisioningClientHandle->_internal.ulProvisioningSymmetricKeyLength = ulSymmetricKeyLength;
+        xAzureIoTProvisioningClientHandle->_internal.pxProvisioningTokenRefresh = prvAzureIoTProvisioningClientTokenGet;
+        xAzureIoTProvisioningClientHandle->_internal.xProvisioningHMACFunction = xHmacFunction;
+        xReturn = eAzureIoTProvisioningSuccess;
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTProvisioningClientResult_t AzureIoTProvisioningClient_Register( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle,
```

Comment:
Not directly related to this but more doxygen for `AzureIoTProvisioningClient_Register`, but... I think we need to be really clear about what a StatusSuccess means here.

It means that the Hub has acknowledged our request and NOT given us some sort of pending code.  But it does NOT mean that we're happy campers, because in the SDK we don't parse out 200/300/400/500 HTTP status codes.  So doxygen saying `After a successful result, AzureIoTProvisioningClient_HubGet() can be called to get the IoT Hub and device ID.`, I really think the flow needs to be like (with actual word-smithing)

"After successful result, your app needs to call AzureIoTProvisioningClient_GetExtendedCode and if it has a happy (200-299??) HTTP status code for top 3 digits, then and only then can it call GetHub().  Otherwise the register call has completed but not succeeded in terms of having device registered"

We could also potentially have the SDK have opinions and not return "statusSuccess" if a 500 came from hub with no retry guidance?  That would be a more natural way I'd think API would work fwiw.

Reply:
And in any case note that [here](https://github.com/Azure/azure-iot-middleware-freertos/blob/f5e07fba53698f26ce03d172ce51134395ad8d59/source/include/azure_iot_provisioning_client.h#L153) the doxygen is calling `AzureIoTProvisioningClient_HubGet` but really should be `AzureIoTProvisioningClient_GetHub`.

Reply:
Currently if the registration on DPS side does not end to Assigned state and there is no retry, we move client side state to ServerError :

[https://github.com/Azure/azure-iot-middleware-freertos/blob/0a198e3f94adb2f557ddca9118cd0692cd0b87bf/source/azure_iot_provisioning_client.c#L424](https://github.com/Azure/azure-iot-middleware-freertos/blob/0a198e3f94adb2f557ddca9118cd0692cd0b87bf/source/azure_iot_provisioning_client.c#L424)

In that case return of this function should give success only in assigned state of DPS.

---
In reply to: [603559700](https://github.com/Azure/azure-iot-middleware-freertos/pull/49#discussion_r603559700) [](ancestors = 603559700)